Mopy/bash/brec.py

```diff
@@ -136,14 +139,179 @@ def pack(self):
+    """
+
+    # size = NewRecordHeader.get_magic_number()
+
+    pack_formats = {
+        'RECORD': ('=4s5I', '=4s4I'),
+        'GRUP0' : ('=4sI4s3I', '=4sI4s2I'),
+        'GRUP1' : ('=4s5I', '=4s4I'),
+        'GRUP2' : ('=4sIi3I', '=4sIi2I'),
+        'GRUP3' : ('=4sIi3I', '=4sIi2I'),
+        'GRUP4' : ('=4sI2h3I', '=4sI2h2I'),
+        'GRUP5' : ('=4sI2h3I', '=4sI2h2I'),
+        'GRUP6' : ('=4s5I', '=4s4I'),
+        'GRUP7' : ('=4s5I', '=4s4I'),
+        'GRUP8' : ('=4s5I', '=4s4I'),
+        'GRUP9' : ('=4s5I', '=4s4I'),
+        'GRUP10': ('=4s5I', '=4s4I'),
+    }  # use 1 through 10 for 'GRUP' + str(groupType)
+
+    @staticmethod
+    def is_top_group(rec_type):
+        """ Return true when the record type is a top group."""
+        return rec_type == 'GRUP'
+
+    def __init__(self,recType='TES4',size=0,arg1=0,arg2=0,arg3=0,extra=0):
+        """
+        :param recType: signature of record
+                      : For GRUP this is always GRUP
+                      : For Records this will be TES4, GMST, KYWD, etc
+        :param size : size of current record, not entire file
+        :param arg1 : For GRUP type of records to follow, GMST, KYWD, etc
+                    : For Records this is the record flags
+        :param arg2 : For GRUP Group Type 0 to 10 see UESP Wiki
+                    : Record FormID, TES4 records have FormID of 0
+        :param arg3 : For GRUP 2H = possible time stamp, unknown
+                    : Record possible version control in CK
+        :param extra: For GRUP 2H, form_version, unknown
+                    : For Records 2H, form_version, unknown
+        """
+        self.recType = recType
+        self.size = size
+        if self.is_top_group(recType):
+            self.label = arg1
+            self.groupType = arg2
+            self.stamp = arg3
+        else:
+            self.flags1 = arg1
+            self.fid = arg2
+            self.flags2 = arg3
+        self.extra = extra
+
+    @staticmethod
+    def get_unique_label(group_type_or_form, label_or_int):
+        """ Unpack current label_or_int and depending on the Group Type
+        pack label_or_int with the correct cast type. Such as 4 chars,
+        signed integer or two shorts.  Currently the two shorts are not used
+        because in other places of the code, it takes 4 bytes and unpacks it
+        to use as coordinates for the cells.
+        """
+        temp_var = label_or_int
+        if group_type_or_form == 0:
+            temp_var = struct.unpack('=4s', struct.pack('=I',label_or_int))[0]
+        if group_type_or_form in [2, 3]:
+            temp_var = struct.unpack('=i', struct.pack('=I',label_or_int))[0]
+        # if group_type_or_form in [4, 5]:
+        #     temp_var_1, temp_var_2 = struct.unpack('=2h', struct.pack('=I',label_or_int))
+        #     temp_var = (temp_var_1, temp_var_2,)
+        return temp_var
+
+    @classmethod
+    def set_magic_number(cls, current_game):
```

Comment:
If refactored as above this not needed - you just use (bush.)game.pack_formats and (bush.)game.header_size (yes thios has also to be defined in game files)